### PR TITLE
mgr: 'osd df' by specified class or (crush) name

### PR DESCRIPTION
**Pull request by xiexingguo**
For large clusters, we use device classes to isolate storage pools.
The existing 'osd df' output turns out to be too nosiy, say, if
you care about only single storage pool with osds possibly spanning over
all hosts.

With this change you are now being able to do 'osd df' by class (or by pool,
if you simply use classes to separate different pools), or by a specified
crush bucket name you are currently interested in, which is much more
convenient.

Some examples:
```
$ bin/ceph osd df tree
ID CLASS WEIGHT  REWEIGHT SIZE   RAW USE DATA    OMAP META  AVAIL   %USE  VAR  PGS STATUS TYPE NAME
-1       0.05878        - 60 GiB 6.4 GiB  23 MiB  0 B 6 GiB  54 GiB 10.60 1.00   -        root default
-3       0.02939        - 30 GiB 3.2 GiB  12 MiB  0 B 3 GiB  27 GiB 10.60 1.00   -            host ceph11
 3   aaa 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  56     up         osd.3
 4   bbb 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  58     up         osd.4
 5   ccc 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  60     up         osd.5
-5       0.02939        - 30 GiB 3.2 GiB  12 MiB  0 B 3 GiB  27 GiB 10.60 1.00   -            host ceph12
 0   aaa 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  50     up         osd.0
 1   bbb 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  61     up         osd.1
 2   ccc 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  51     up         osd.2
                    TOTAL 60 GiB 6.4 GiB  23 MiB  0 B 6 GiB  54 GiB 10.60
MIN/MAX VAR: 1.00/1.00  STDDEV: 0

$ bin/ceph osd df tree class aaa
ID CLASS WEIGHT  REWEIGHT SIZE   RAW USE DATA    OMAP META  AVAIL   %USE  VAR  PGS STATUS TYPE NAME
-1       0.05878        - 20 GiB 2.1 GiB 7.8 MiB  0 B 2 GiB  18 GiB 10.60 1.00   -        root default
-3       0.02939        - 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00   -            host ceph11
 3   aaa 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  56     up         osd.3
-5       0.02939        - 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00   -            host ceph12
 0   aaa 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  50     up         osd.0
                    TOTAL 20 GiB 2.1 GiB 7.8 MiB  0 B 2 GiB  18 GiB 10.60
MIN/MAX VAR: 1.00/1.00  STDDEV: 0

$ bin/ceph osd df tree name ceph11
ID CLASS WEIGHT  REWEIGHT SIZE   RAW USE DATA    OMAP META  AVAIL   %USE  VAR  PGS STATUS TYPE NAME
-3       0.02939        - 30 GiB 3.2 GiB  12 MiB  0 B 3 GiB  27 GiB 10.60 1.00   -            host ceph11
 3   aaa 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  56     up         osd.3
 4   bbb 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  58     up         osd.4
 5   ccc 0.00980  1.00000 10 GiB 1.1 GiB 3.9 MiB  0 B 1 GiB 9.0 GiB 10.60 1.00  60     up         osd.5
                    TOTAL 30 GiB 3.2 GiB  12 MiB  0 B 3 GiB  27 GiB 10.60
MIN/MAX VAR: 1.00/1.00  STDDEV: 0

```

Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by xiexingguo** [comment]
retest this please

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -5007,15 +5051,16 @@ class OSDUtilizationPlainDumper : public OSDUtilizationDumper<TextTable> {
...
-	 << byte_u_t(pgmap.get_osd_sum().statfs.allocated)
-	 << byte_u_t(pgmap.get_osd_sum().statfs.omap_allocated)
-	 << byte_u_t(pgmap.get_osd_sum().statfs.internal_metadata)
-	 << byte_u_t(pgmap.get_osd_sum().statfs.available)
+	 << byte_u_t(pgmap.get_osd_sum(dumped).statfs.total)
+	 << byte_u_t(pgmap.get_osd_sum(dumped).statfs.get_used_raw())
+	 << byte_u_t(pgmap.get_osd_sum(dumped).statfs.allocated)
+	 << byte_u_t(pgmap.get_osd_sum(dumped).statfs.omap_allocated)
+	 << byte_u_t(pgmap.get_osd_sum(dumped).statfs.internal_metadata)
+	 << byte_u_t(pgmap.get_osd_sum(dumped).statfs.available)
```
Comment (liewegas) [inline src/osd/OSDMap.cc]: instead of calling get_osd_sum() 6 times, let's call it once and put it in temp variable?

File: src/osd/OSDMap.cc [hunk 2]
```diff
@@ -4815,19 +4815,51 @@ class OSDUtilizationDumper : public CrushTreeDumper::Dumper<F> {
...
   }
 
 protected:
+
+  bool should_dump(int id) const {
+    if (!allowed.empty() && !allowed.count(id)) // filter by name
+      return false;
+    if (id >= 0 && !class_name.empty()) {
+      auto item_class_name = osdmap->crush->get_item_class(id);
+      if (!item_class_name || // not bound to a class yet
```
Comment (liewegas) [inline src/osd/OSDMap.cc]: is operator! on std::string equivalent to empty() ?

**Comment by liewegas** [comment]
I really wish our argparse code was a bit more normal/robust so that we could have option named arguments so that we could do, for example, ``ceph osd df --class foo`` or ``ceph osd df --root default``.  We have a ton of commands that don't follow the usual unix conventions (e.g., the optional ``detail`` argument for ``ceph df``) and it's annoying. :(

File: src/osd/OSDMap.cc (hunk 2 above)
Comment (xiexingguo) [reply src/osd/OSDMap.cc]: `get_item_class` returns `a const char *`, not a string

File: src/osd/OSDMap.cc (hunk 1 above)
Comment (xiexingguo) [reply src/osd/OSDMap.cc]: @liewegas Updated!

File: src/osd/OSDMap.cc (hunk 2 above)
Comment (tchaikov) [reply src/osd/OSDMap.cc]: probably `s/auto/const char */` for better readability?

**Comment by xiexingguo** [comment]
cephtool: http://pulpito.ceph.com/xxg-2019-03-07_03:25:52-rados-wip-osd-df-by-class-distro-basic-smithi/
standalone: http://pulpito.ceph.com/xxg-2019-03-07_03:26:13-rados-wip-osd-df-by-class-distro-basic-smithi/
rados: http://pulpito.ceph.com/xxg-2019-03-07_03:26:35-rados-wip-osd-df-by-class-distro-basic-smithi/

QA looks good.
Can we have this in nautilus? I wouldn't worry about the risk since these are pure CLI enhancements..

**Pull request merged**